src/core/lib/slice/slice_buffer.cc

```diff
@@ -262,54 +276,79 @@ void grpc_slice_buffer_move_into(grpc_slice_buffer* src,
   src->length = 0;
 }
 
+template <bool incref>
+static void slice_buffer_move_single_slice(grpc_slice_buffer* src,
+                                           grpc_slice_buffer* dst,
+                                           size_t bytes_extra) {
+  // Last slice in batch for slice_buffer_move_first_maybe_ref.
+  // We may have to consume it  whole, or just a chunk.
+  if (bytes_extra) {
+    // Just a chunk.
+    if (incref) {
+      dst->slices[dst->count] = grpc_slice_split_head(
+          &src->slices[0], GRPC_SLICE_LENGTH(src->slices[0]) - bytes_extra);
+    } else {
+      dst->slices[dst->count] = grpc_slice_split_head_noref(
+          &src->slices[0], GRPC_SLICE_LENGTH(src->slices[0]) - bytes_extra);
+    }
+  } else {
+    // The whole thing.
+    dst->slices[dst->count] = src->slices[0];
+    ++src->slices;
+    --src->count;
+  }
+  ++dst->count;
+}
+
+template <bool incref>
 static void slice_buffer_move_first_maybe_ref(grpc_slice_buffer* src, size_t n,
-                                              grpc_slice_buffer* dst,
-                                              bool incref) {
-  GPR_ASSERT(src->length >= n);
-  if (src->length == n) {
+                                              grpc_slice_buffer* dst) {
+  if (n == 0) {
+    return;
+  }
+  if (n == src->length) {
     grpc_slice_buffer_move_into(src, dst);
     return;
   }
+  GPR_ASSERT(src->length > n);
+
+  size_t slices_to_copy = 0;
+  size_t bytes_so_far = 0;
+  size_t curr_idx = 0;
+  while (bytes_so_far < n) {
+    bytes_so_far += GRPC_SLICE_LENGTH(src->slices[curr_idx]);
+    ++slices_to_copy;
+    ++curr_idx;
```

Comment:
Although the style guide allows either, our codebase tends to use post-increment when we don't care about the value. Can this be changed for a (possibly foolish) consistency?

Reply:
Although I like how this avoids the use of break later on, is there a way to avoid having a pre-loop before an implicit loop (memcpy)? Perhaps there is no benefit of doing so. Either way, can I suggest a comment before the loop to indicate what you're accomplishing, so that we don't lose track of it in the future?

Reply:
Regarding pre-looping: the main thought here is that since the last slice may need to be split, then in the one-slice-at-a-time case we have to check for each slice whether it needs to be split (i.e. len_so_far = 0, while len_so_far < n, currlen = slice.len, len_so_far += currlen, if len_so_far > n then split slice and add the split). So per iteration, we have two comparisons.

If we preloop: (i.e. len_so_far = 0, while len_so_far < n, currlen = slice.len, len_so_far += currlen) we only need to check if we need to split that last slice once. So for n=100 slices, that's 100 cmp total rather than 200 if we go slice by slice.

Reply:
Makes sense to me.